src/notary/mod.rs

```diff
@@ -1,24 +1,118 @@
+use std::sync::Arc;
+
+use http_body_util::{BodyExt as _, Either, Empty, Full};
+use hyper::{client::conn::http1::Parts, Request, StatusCode};
+use hyper_util::rt::TokioIo;
+use notary_server::{ClientType, NotarizationSessionRequest, NotarizationSessionResponse};
+use rustls::{Certificate, ClientConfig, RootCertStore};
+use tokio::net::TcpStream;
+use tokio_rustls::TlsConnector;
+use tokio_util::bytes::Bytes;
+
+/// Requests notarization from the Notary server.
+pub async fn request_notarization(
+    host: &str,
+    port: u16,
+    max_sent_data: Option<usize>,
+    max_recv_data: Option<usize>,
+) -> (tokio_rustls::client::TlsStream<TcpStream>, String) {
+    // Connect to the Notary via TLS-TCP
+    let pem_file = std::str::from_utf8(include_bytes!(
+        "../fixture/rootCA.crt"
+    ))
+    .unwrap();
+    let mut reader = std::io::BufReader::new(pem_file.as_bytes());
+    let mut certificates: Vec<Certificate> = rustls_pemfile::certs(&mut reader)
+        .unwrap()
+        .into_iter()
+        .map(Certificate)
+        .collect();
+    let certificate = certificates.remove(0);
+
+    let mut root_store = RootCertStore::empty();
+    root_store.add(&certificate).unwrap();
+
+    let client_notary_config = ClientConfig::builder()
+        .with_safe_defaults()
+        .with_root_certificates(root_store)
+        .with_no_client_auth();
+    let notary_connector = TlsConnector::from(Arc::new(client_notary_config));
+
+    let notary_socket = tokio::net::TcpStream::connect((host, port)).await.unwrap();
+
+    let notary_tls_socket = notary_connector
+        // Require the domain name of notary server to be the same as that in the server cert
+        .connect("tlsnotaryserver.io".try_into().unwrap(), notary_socket)
         .await
         .unwrap();
+
+    // Attach the hyper HTTP client to the notary TLS connection to send request to the /session endpoint to configure notarization and obtain session id
+    let (mut request_sender, connection) =
+        hyper::client::conn::http1::handshake(TokioIo::new(notary_tls_socket))
+            .await
+            .unwrap();
+
+    // Spawn the HTTP task to be run concurrently
+    let connection_task = tokio::spawn(connection.without_shutdown());
+
+    // Build the HTTP request to configure notarization
+    let payload = serde_json::to_string(&NotarizationSessionRequest {
+        client_type: ClientType::Tcp,
+        max_sent_data,
+        max_recv_data,
+    })
+    .unwrap();
+
+    let request = Request::builder()
+        .uri(format!("https://{host}:{port}/session"))
+        .method("POST")
+        .header("Host", host)
+        // Need to specify application/json for axum to parse it as json
+        .header("Content-Type", "application/json")
+        .body(Either::Left(Full::new(Bytes::from(payload))))
+        .unwrap();
+
+    let configuration_response = request_sender.send_request(request).await.unwrap();
+    assert!(configuration_response.status() == StatusCode::OK);
```

Comment:
`assert` will panic when the condition is false, please use `Err` (from `Result`) so we can handle the failure

Reply:
Changed in latest commit, now `request_notarization` will return error in case of failure